### PR TITLE
US075-AddTrailerOffer

### PR DESCRIPTION
**Pull request by JoJ3o**
[Aanhangers & aanbod toevoegen](https://trello.com/c/y3wnqTHD/75-aanhangers-aanbod-toevoegen)

- Add TrailerOffer POST Route
- Connect owner to trailerOffer
- TrailerOffer Data validation
- Move Authentication logic to TrailerOfferService
- Create custom TrailerTypeNotFoundException
- Validate date of all TrailerOffer fields
- Convert TrailerOffer requestBody to entity
- Separate TrailerOfferDTO to TrailerInfoDTO (homepage display) and CreateTrailerOfferDTO (convert data to entity)
- Catch customer and trailerType exceptions
- Add optional keyword in TrailerTypeRepository
- Turn DTO classes into records
- Add JsonAlias for all camelCase fields

### CODE REVIEW

**Comment by JoJ3o** [comment]
[User Story Experimenting](https://trello.com/c/6pADZYZb/130-user-story-experimenting)

**Comment by JoJ3o** [comment]
[Aanhangers & aanbod toevoegen](https://trello.com/c/y3wnqTHD/75-aanhangers-aanbod-toevoegen)

File: api/src/main/java/com/buurbak/api/trailers/controller/TrailerOfferController.java [hunk 1]
```diff
@@ -24,9 +32,28 @@ public TrailerOffer getTrailerOffer(@PathVariable UUID id){
...
+            @ApiResponse(responseCode = "201", description = "Created"),
+            @ApiResponse(responseCode = "400", description = "Bad Request", content = @Content),
+            @ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content),
+            @ApiResponse(responseCode = "404", description = "Entity not found", content = @Content)
+    })
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public void addTrailerOffer(@Valid @RequestBody CreateTrailerOfferDTO createTrailerOfferDTO, Authentication authentication) {
+        try {
+            trailerOfferService.addTrailerOffer(createTrailerOfferDTO, authentication.getName());
```
Comment (spark-156) [inline api/src/main/java/com/buurbak/api/trailers/controller/TrailerOfferController.java:50]: return ID?

File: api/src/main/java/com/buurbak/api/trailers/model/TrailerOffer.java [hunk 2]
```diff
@@ -20,14 +21,13 @@ public class TrailerOffer {
     @GeneratedValue
     private UUID id;
 
-    @ManyToOne(fetch = FetchType.EAGER)
+    @ManyToOne
```
Comment (spark-156) [inline api/src/main/java/com/buurbak/api/trailers/model/TrailerOffer.java:24]:
```suggestion
    @ManyToOne(fetch = FetchType.EAGER)
```

File: api/src/main/java/com/buurbak/api/trailers/controller/TrailerOfferController.java (hunk 1 above)
Comment (spark-156) [inline api/src/main/java/com/buurbak/api/trailers/controller/TrailerOfferController.java:50]:
```suggestion
    public String addTrailerOffer(@Valid @RequestBody CreateTrailerOfferDTO createTrailerOfferDTO, Authentication authentication) {
        try {
            TrailerOffer trailerOffer = trailerOfferService.addTrailerOffer(createTrailerOfferDTO, authentication.getName());
            return trailerOffer.getId().toString();
```

File: api/src/main/java/com/buurbak/api/trailers/service/TrailerOfferService.java [hunk 3]
```diff
@@ -13,13 +19,38 @@
...
+                createTrailerOfferDTO.licensePlateNumber(),
+                createTrailerOfferDTO.pickUpTimeStart(),
+                createTrailerOfferDTO.pickUpTimeEnd(),
+                createTrailerOfferDTO.dropOffTimeStart(),
+                createTrailerOfferDTO.dropOffTimeEnd(),
+                createTrailerOfferDTO.location(),
+                createTrailerOfferDTO.price(),
+                createTrailerOfferDTO.available());
+        trailerOfferRepository.save(trailerOffer);
+    }
```
Comment (spark-156) [inline api/src/main/java/com/buurbak/api/trailers/service/TrailerOfferService.java:55]:
```suggestion
    public TrailerOffer addTrailerOffer(CreateTrailerOfferDTO createTrailerOfferDTO, String username) throws CustomerNotFoundException, TrailerTypeNotFoundException {
        Customer customer = customerService.findByUsername(username);
        TrailerType trailerType = trailerTypeService.findByName(createTrailerOfferDTO.trailerType());
        TrailerOffer trailerOffer = new TrailerOffer(
                trailerType,
                customer,
                createTrailerOfferDTO.length(),
                createTrailerOfferDTO.height(),
                createTrailerOfferDTO.width(),
                createTrailerOfferDTO.weight(),
                createTrailerOfferDTO.capacity(),
                createTrailerOfferDTO.licensePlateNumber(),
                createTrailerOfferDTO.pickUpTimeStart(),
                createTrailerOfferDTO.pickUpTimeEnd(),
                createTrailerOfferDTO.dropOffTimeStart(),
                createTrailerOfferDTO.dropOffTimeEnd(),
                createTrailerOfferDTO.location(),
                createTrailerOfferDTO.price(),
                createTrailerOfferDTO.available());
        trailerOfferRepository.save(trailerOffer);
        return trailerOffer;
    }
```

**Review by spark-156 — CHANGES_REQUESTED** [review]

**Review by spark-156 — APPROVED** [review]

**Pull request merged**